File: crates/flighthq-scene-wgpu/src/wireframe_wgpu_mesh_material_renderer.rs
//! The built-in Wireframe forward mesh-material renderer — the WGSL mirror of
//! `wireframe_gl_mesh_material_renderer`.
//!
//! Ports `@flighthq/scene-wgpu` `wireframeWgpuMeshMaterialRenderer.ts`. `bind` selects
//! the line-list wireframe pipeline for the color format, writes the shared Frame
//! uniform (lights ignored), and binds the flat linear line color at group(2).
//! `thickness` > 1 is not honored (WebGPU has no line-width control).
//!
//! TS↔Rust divergence: the TS renderer draws the mesh's derived line-index buffer
//! (`wgpuWireframeUpload`) as line-list primitives. That derived line-index upload is
//! not yet ported to the Rust crate (mirroring the `wireframe_gl` sibling's stubbed
//! line-index upload), so `draw` here issues the shared indexed draw over the
//! triangle index buffer under the line-list pipeline — `bind` is full; the true
//! edge topology arrives with the ported wireframe upload.

use flighthq_materials::wireframe_material_kind;
use flighthq_render_wgpu::WgpuRenderState;
use flighthq_types::camera::Camera;
use flighthq_types::material::Material;
use flighthq_types::scene_render::{SceneLightBlock, SceneRenderProxy};
use flighthq_types::unlit_material::WireframeMaterial;

use crate::wgpu_mesh_material_registry::{
    WgpuMeshMaterialRenderer, register_wgpu_mesh_material_renderer,
};
use crate::wgpu_mesh_pipeline::{
    draw_wgpu_mesh_material_subset, unpack_color_to_linear_f32, write_wgpu_frame_uniform,
};
use crate::wgpu_scene_runtime::{WgpuMeshUpload, WgpuSceneRuntime};
use crate::wgpu_wireframe_prelude::{bind_wgpu_wireframe_color, ensure_wgpu_wireframe_pipeline};

/// The built-in Wireframe forward mesh-material renderer.
pub struct WireframeWgpuMeshMaterialRenderer;

impl WgpuMeshMaterialRenderer for WireframeWgpuMeshMaterialRenderer {
    fn bind(
        &self,
        state: &mut WgpuRenderState,
        scene: &mut WgpuSceneRuntime,
        material: Option<&dyn Material>,
        _lights: &SceneLightBlock,
        camera: &Camera,
    ) {
        let wireframe = material.and_then(downcast_wireframe);
        let format = state.runtime.current_color_format.unwrap_or(state.format);
        let cache_key = ensure_wgpu_wireframe_pipeline(state, scene, format);
        scene.active_pipeline_key = Some(cache_key.clone());
        write_wgpu_frame_uniform(state, scene, camera, _lights);

        let material_key = material.map_or_else(wireframe_material_kind, |m| m.kind());
        let color = wireframe.map_or(WHITE, |m| unpack_color_to_linear_f32(m.color));
        bind_wgpu_wireframe_color(state, scene, &cache_key, material_key, &color);
        scene.active_material_key = Some(material_key);
    }

    fn draw(
        &self,
        state: &mut WgpuRenderState,
        scene: &mut WgpuSceneRuntime,
        proxy: &SceneRenderProxy,
        upload: &WgpuMeshUpload,
    ) {
        draw_wgpu_mesh_material_subset(state, scene, proxy, upload);
    }
}

/// Returns the singleton Wireframe renderer instance.
pub fn wireframe_wgpu_mesh_material_renderer() -> WireframeWgpuMeshMaterialRenderer {
    WireframeWgpuMeshMaterialRenderer
}

/// Registers the built-in Wireframe renderer for the Wireframe material kind on this
/// scene runtime. Opt-in (no top-level side effect).
pub fn register_wireframe_wgpu_material(state: &mut WgpuRenderState, scene: &mut WgpuSceneRuntime) {
    register_wgpu_mesh_material_renderer(
        state,
        scene,
        wireframe_material_kind(),
        Box::new(WireframeWgpuMeshMaterialRenderer),
    );
}

fn downcast_wireframe(material: &dyn Material) -> Option<&WireframeMaterial> {
    (material as &dyn core::any::Any).downcast_ref::<WireframeMaterial>()
}

const WHITE: [f32; 4] = [1.0, 1.0, 1.0, 1.0];

#[cfg(test)]
mod tests {
    use super::*;
    use crate::wgpu_scene_runtime::create_wgpu_scene_runtime;

    mod register_wireframe_wgpu_material {
        use super::*;

        #[test]
        fn registers_the_renderer_under_the_wireframe_kind() {
            let mut scene = create_wgpu_scene_runtime();
            scene.material_registry.insert(
                wireframe_material_kind(),
                Box::new(WireframeWgpuMeshMaterialRenderer),
            );
            assert!(
                scene
                    .material_registry
                    .contains_key(&wireframe_material_kind())
            );
        }
    }
}

File: packages/scene-wgpu/src/wireframeWgpuMeshMaterialRenderer.ts
import type { LinearColor } from '@flighthq/color';
import { unpackColorToLinear } from '@flighthq/color';
import { getWgpuRenderStateRuntime } from '@flighthq/render-wgpu';
import type {
  Camera,
  Material,
  MeshGeometry,
  SceneLightBlock,
  SceneRenderProxy,
  WgpuMeshMaterialRenderer,
  WgpuRenderState,
  WireframeMaterial,
} from '@flighthq/types';
import { WireframeMaterialKind } from '@flighthq/types';

import { registerWgpuMeshMaterialRenderer } from './wgpuMeshMaterialRegistry';
import { beginWgpuMeshDraw, writeWgpuDrawUniform, writeWgpuFrameUniform } from './wgpuMeshPipeline';
import { getWgpuSceneRuntime } from './wgpuSceneRuntime';
import { bindWgpuWireframeColor, ensureWgpuWireframePipeline } from './wgpuWireframePrelude';
import { ensureWgpuWireframeUpload } from './wgpuWireframeUpload';

// The built-in Wireframe forward renderer (WgpuMeshMaterialRenderer for WireframeMaterialKind) — the
// WGSL mirror of wireframeGlMeshMaterialRenderer. Draws the mesh's triangle edges as line-list
// primitives in a single flat linear color. Unlike the triangle families it does not use
// drawWgpuMeshSubset: draw binds the derived line-index buffer (see wgpuWireframeUpload) and issues a
// line-list indexed draw over the subset's line range. `thickness` > 1 is not honored (WebGPU has no
// line-width control); the field is documented as best-effort and ignored. Lights are ignored. See
// registerWireframeWgpuMaterial to install it.
export const wireframeWgpuMeshMaterialRenderer: WgpuMeshMaterialRenderer = {
  bind(
    state: WgpuRenderState,
    material: Readonly<Material> | null,
    _lights: Readonly<SceneLightBlock>,
    camera: Readonly<Camera>,
  ): void {
    const stateRuntime = getWgpuRenderStateRuntime(state);
    const pass = stateRuntime.renderPass;
    if (pass === null) return;

    const wireframe = material as Readonly<WireframeMaterial> | null;
    const format = stateRuntime.currentColorFormat ?? state.format;
    const pipeline = ensureWgpuWireframePipeline(state, format);
    writeWgpuFrameUniform(state, camera, _lights);

    let group: GPUBindGroup;
    if (wireframe === null) {
      group = bindWgpuWireframeColor(state, pipeline, FALLBACK_MATERIAL, WHITE);
    } else {
      unpackColorToLinear(_scratch, wireframe.color);
      group = bindWgpuWireframeColor(state, pipeline, wireframe, _scratch);
    }

    beginWgpuMeshDraw(state, pipeline);
    pass.setBindGroup(2, group);
  },

  draw(state: WgpuRenderState, proxy: Readonly<SceneRenderProxy>, geometry: Readonly<MeshGeometry>): void {
    const stateRuntime = getWgpuRenderStateRuntime(state);
    const pass = stateRuntime.renderPass;
    const scene = getWgpuSceneRuntime(state);
    if (pass === null || scene.activeMeshPipeline === null) return;

    const subset = proxy.subset;
    if (subset.indexCount === 0) return;

    const upload = ensureWgpuWireframeUpload(state, geometry);
    if (upload === null) return;

    const drawBindGroup = writeWgpuDrawUniform(state, proxy);
    _dynamicOffsets[0] = scene.pendingDrawOffset;

    pass.setBindGroup(1, drawBindGroup, _dynamicOffsets);
    pass.setVertexBuffer(0, upload.vertexBuffer);
    pass.setIndexBuffer(upload.lineIndexBuffer, upload.indexFormat);
    // Each triangle index contributes two line indices, so the subset's line range is its triangle
    // range scaled by 2.
    pass.drawIndexed(subset.indexCount * 2, 1, subset.indexOffset * 2, 0, 0);
  },
};

// Registers the built-in Wireframe renderer for WireframeMaterialKind on this state. Opt-in (no top-
// level side effect); call once per WgpuRenderState before drawWgpuScene so meshes with
// WireframeMaterials draw.
export function registerWireframeWgpuMaterial(state: WgpuRenderState): void {
  registerWgpuMeshMaterialRenderer(state, WireframeMaterialKind, wireframeWgpuMeshMaterialRenderer);
}

const _scratch: LinearColor = [0, 0, 0, 0];
const _dynamicOffsets = new Uint32Array(1);
const WHITE: LinearColor = [1, 1, 1, 1];
const FALLBACK_MATERIAL = {} as Readonly<WireframeMaterial>;
